### PR TITLE
Patch 2019-09-30: New characters, weapons, uncaps

### PR DESCRIPTION
**Pull request by kei-gbf**
# Checklist

## New Character
- [x] [Siegfried (Fire)](https://gbf.wiki/Siegfried_(Fire))
- [x] [Monika (Grand)](https://gbf.wiki/Monika_(Grand))
- [x] [La Coiffe (SR)](https://gbf.wiki/La_Coiffe_(SR)) 

## New Weaponss
- [x] [Black Dragon's Maw](https://gbf.wiki/Black_Dragon%27s_Maw)
- [x] [Sky Ace](https://gbf.wiki/Sky_Ace)
- [x] 4 stars Sky Ace
- [Lilac Comb](https://gbf.wiki/Lilac_Comb) no need implement SR weapon

## New Summons
- [Princess Long Ji](https://gbf.wiki/Princess_Long_Ji) no need implement summon detail

## Uncaps
- [x] [Rackam (Grand)](https://gbf.wiki/Rackam_(Grand))
- [x] [Katalina (Grand)](https://gbf.wiki/Katalina_(Grand))

## Update scripts

- [x] chara convert script
  - [x] support "/" separator for "fav1/fav2" 
  - ~~support "_full" suffix for image URL file name~~
- download script
  - ~~transform "_?full" suffix url for game url (remove _?full)~~ 

----
## Tasks (Done)
- [x] Update txt soruce and translation
  - [x] add min hp/atk manually (wiki missing)
  - [x] update lv100 status for uncap characters
- [x] run converter script

## Questions
- [x] where should insert new chara/arm record in txt file ?
- [x] need update uncap character image ?

### CODE REVIEW

**Comment by Groovman** [comment]
Sky Ace is a Grand Weapon and is a katana so it goes after Ichigo because higher ID,.
Black Dragon's Maw is a Normal Weapon and is a katana, so it comes after the last normal weapon katana which should be just before the 100 goldmoon weapon start.

Characters are sorted in the following way, Eternals, Arcanum, SSR (newest first)m SR (oldest first)
Monika (Grand) after Nier (last ID of Arcanum characters)
Siegfried (Fire) after Monika (Grand) 
La Coiffe (SR) goes to the end of the character list

> no need implement SR weapon ?
No people would not use 99.9% of them anyway in a grid especially normal weapons (sr seraph weapons are the exception, or magna ones), and they are at a state where they don't need to min-max their grids with sr weapons anyway.

> need update uncap character image ?
would do it so people know they have been updated to their 5* stats.

**Comment by kei-gbf** [comment]
Thanks advice

- [x] Sky Ace is a Grand Weapon and is a katana so it goes after Ichigo
- [x] Black Dragon's Maw is *snip* before the 100 goldmoon weapon start.
- [x] moved Monika (Grand) and Siegfried (Fire) after Neir  (order: Neir, Monika, Siegfried)
- [x] La Coiffe (SR) stay end of file
- [x] Update 5* chara image
- [x] no SR weapon ... ok

File: charaData.json [hunk 1]
```diff
@@ -659,6 +659,50 @@
...
     },
+    "モニカ(リミテッドver)": {
+        "name": "モニカ(リミテッドver)",
+        "ja": "モニカ(リミテッドver)",
+        "element": "wind",
+        "type": "balance",
+        "race": "human",
+        "sex": "female",
+        "fav1": "sword",
+        "fav2": "none",
```
Comment (kagekiyo7) [inline charaData.json]: Monika's fav2 is katana.
https://i0.wp.com/tkgranblue.com/wp-content/uploads/2019/08/2019-09-30-1.jpg

File: charaData.json [hunk 2]
```diff
@@ -659,6 +659,50 @@
...
+    },
+    "ジークフリート(火属性ver)": {
+        "name": "ジークフリート(火属性ver)",
+        "ja": "ジークフリート(火属性ver)",
+        "element": "fire",
+        "type": "pecu",
+        "race": "human",
+        "sex": "male",
+        "fav1": "sword",
+        "fav2": "none",
```
Comment (kagekiyo7) [inline charaData.json]: Siegfried's fav2 is also katana.
https://livedoor.blogimg.jp/hecate_mk2-gbf/imgs/e/5/e5b82a62.jpg

File: charaData.json [hunk 3]
```diff
@@ -4192,13 +4236,13 @@
...
+        "hp": "1900",
         "minattack": "1500",
-        "attack": "8000",
+        "attack": "9500",
         "baseDA": 10.0,
         "baseTA": 5.0,
-        "ougiRatio": 4.5,
-        "imageURL": "./charaimgs/3040059000_01.png",
+        "ougiRatio": 5.0,
+        "imageURL": "./charaimgs/3040060000_01.png",
```
Comment (kagekiyo7) [inline charaData.json]: Is imageURL correct?

File: charaData.json [hunk 4]
```diff
@@ -4302,13 +4346,13 @@
...
+        "hp": "2310",
         "minattack": "1350",
-        "attack": "8100",
+        "attack": "9450",
         "baseDA": 7.0,
         "baseTA": 3.0,
-        "ougiRatio": 4.5,
-        "imageURL": "./charaimgs/3040054000_01.png",
+        "ougiRatio": 5.0,
+        "imageURL": "./charaimgs/3040055000_91.png",
```
Comment (kagekiyo7) [inline charaData.json]: Is imageURL correct?

**Comment by kei-gbf** [comment]
```python
br_pattern = re.compile("(\w+)&br;(\w+)")
```

txt_source had "剣/刀" but the converter script requires "fav1&br;fav2" format.
I will modify the converter script side. (to remove the manual edit task)

5stars imageURL txt_source ok, it was missing re-generate json file.
ty for review.

**Comment by kei-gbf** [comment]
> 5stars imageURL txt_source ok, it was missing re-generate json file.

no, my misunderstand.
converter script did not catch the new image file format.

```python
key_pattern = re.compile("(\d+_\d+\.png)")
```
but the new image file was `3040054000_03_full`

and found script bugs, when key pattern did not match, use previous row's data.

```python
m = key_pattern.search(row[1])
if m:
    key = m.group(1)
# No else: case ... previous `key` is used.
```

**Comment by kei-gbf** [comment]
> but the new image file was 3040054000_03_full

or I got the file name wrong place?
I could not find another "_full" image files.

**Comment by kagekiyo7** [comment]
Wiki probably got wrong name image.

Inaccessible:
http://gbf.game-a.mbga.jp/assets/img/sp/assets/npc/b/3040059000_03_full.png

Accessible:
http://gbf.game-a.mbga.jp/assets/img/sp/assets/npc/b/3040059000_03.png

**Comment by kagekiyo7** [comment]
I found other uncap character (Vania(ヴァンピィ)) is also same bug.

Accessible:
http://gbf.game-a.mbga.jp/assets/img/sp/assets/npc/b/3040057000_03.png
https://gbf-wiki.com/index.php?plugin=attach&refer=img&openfile=3040057000_03_full.png

Inaccessible:
http://gbf-wiki.com/index.php?plugin=attach&refer=img&openfile=3040057000_03.png

**Comment by kei-gbf** [comment]
re-open #186?

__WARNING__ those links will make a request to game site.

NEED HELP: how to update imageURL for uncapped chara

**Comment by kagekiyo7** [comment]
Idea:
1. use previous character images.
2. stop support either wiki or granblue.
3. add "_full" to the file name in wiki url list.
→ impossible
https://gbf-wiki.com/index.php?plugin=attach&refer=img&openfile=3040057000_03_full.png
https://gbf-wiki.com/index.php?plugin=attach&refer=img&openfile=3040005000_03full.png
4. ignore (mimumemo's motocal probably got game url.)
image

**Comment by Groovman** [comment]
I am for ditching gbf wiki can't download them from there anyway because gaijin :D

**Comment by kei-gbf** [comment]
:+1: Delete "_?full" for game url (no effect to wiki url)

```python
re.sub("_?full.png", ".png", "03_full.png") => "03.png"
re.sub("_?full.png", ".png", "03full.png") => "03.png"
```

download script we have now `transform_wiki_url` function,
I will add transform_game_url function to remove _full suffix.

----
- [ ] Confirm download_image.py script can download the file name with "_full"
     because, transform_wiki_url convert the url to direct link.

**Comment by kei-gbf** [comment]
Ah, I misunderstand again
Should I delete "_full" from txt_source ?

**Comment by Groovman** [comment]
Yes remove the _full from the txt_source. Also I am not sure but some gbf wiki images don't work with _full also, because the change was only for newer ones, so transformation all in batch might also break things again. I don't know if they also changed the older images path.

**Comment by kei-gbf** [comment]
This PR will not change other previous chara imageURL
expect Katalina and Rackam

6b7390c add support "_full" url for wiki/game.

- download script work _full now with transform _full suffix (only transform for game site)
  - ex. game) request to _03.png save as _03_full.png
  - ex. wiki) request to _03_full.png save as _03_full.png
- wget shell script with game url does not work the _full url (no transform)
    but we don't have shell script now. removed from Readme seem it's out of support.

the issue #186 was for the case download from wiki site.
some of uncapped chara imageURL did not have "_?full" suffix which wiki uploaded image has.

so require test game/wiki cases

- [ ] download chara image from wiki site
- [ ] download chara image from game site

**Comment by kei-gbf** [comment]
If those fix _full patch may take a time for compatibility tests.
I will revert 6b7390c and remove _full from txt_source

This PR is game update (higher priority),
not expect to interrupt the update merge. I will make separate PR for the fix.

**Comment by kei-gbf** [comment]
revert 6b7390c ... transform url in download script was not so good approach,
It should be transformed in txt_source/charaImageGameURLList.txt

- [x] keep: br_pattern for "fav1/fav2"
- [x] drop: key_pattern for _full imageURL ... will be another PR

**Comment by kagekiyo7** [comment]
I don't want to test because I'm afraid of BAN…
Could someone use Proxy?

**Comment by kei-gbf** [comment]
the fix patch for game site was dropped.
now the new image url format is same as other uncap chara's
(so it's not found for download from wiki)

- wiki had `_full` suffix `_03_full.png` <- 6b7390c support, but it had problem then reverted
- game site has no suffix, just `_03.png` for uncap chara <-- current head implementation

If other uncap chara also has the same issue? it should be another bugfix PR later.
this PR will be just game update patch. (expect br_pattern fix for `fav1/fav2`)

**Comment by kagekiyo7** [comment]
OK, I had missed.

**Comment by kei-gbf** [comment]
This is not proxy though, low risk way is using cloud sandbox environment
`https://gitpod.io/#https://github.com/MotocalDevelopers/motocal/pull/364`

```
> python3 ./scripts/download_image.py --check chara

or 

> npm run download:chara -- -c
```
will report list of Not found URL (default: download from wiki)
all were files with _03.png, maybe uncap charas.

```
> python3 ./scripts/download_image.py --check --site=game chara
```
was OK (nothing to report then show nothing)

```
> python3 ./scripts/download_image.py --site=wiki chara
> python3 ./scripts/download_image.py --site=game chara
```
1. download files from wiki site (expect uncap chara)
2. download files from game site (download only missing files)

----
```
> npm run download:help
```
for more info download script options.

----
if they block gitpod.io's address then everyone on gitpod.io can't test the download script.
pls do not abuse even via proxy. (just not run many time/often)

**Pull request merged**